src/device/socket/vsock.rs

```diff
@@ -0,0 +1,170 @@
+            let features = Feature::from_bits_truncate(features);
+            info!("Device features: {:?}", features);
+            // negotiate these flags only
+            let supported_features = Feature::empty();
+            (features & supported_features).bits()
+        });
+
+        let config = transport.config_space::<VirtioVsockConfig>()?;
+        info!("config: {:?}", config);
+        // Safe because config is a valid pointer to the device configuration space.
+        let guest_cid = unsafe { volread!(config, guest_cid_low) as u64 };
+        info!("guest cid: {guest_cid:?}");
+
+        let mut rx = VirtQueue::new(&mut transport, RX_QUEUE_IDX)?;
+        let tx = VirtQueue::new(&mut transport, TX_QUEUE_IDX)?;
+        let event = VirtQueue::new(&mut transport, EVENT_QUEUE_IDX)?;
+
+        let queue_buf_dma = Dma::new(1, BufferDirection::DeviceToDriver)?;
+
+        // Safe because no alignment or initialisation is required for [u8], the DMA buffer is
+        // dereferenceable, and the lifetime of the reference matches the lifetime of the DMA buffer
+        // (which we don't otherwise access).
+        let queue_buf_rx = unsafe { queue_buf_dma.raw_slice().as_mut() };
+
+        // Safe because the buffer lives as long as the queue.
+        let _token = unsafe { rx.add(&[], &mut [queue_buf_rx])? };
+
+        if rx.should_notify() {
+            transport.notify(RX_QUEUE_IDX);
+        }
+        transport.finish_init();
+
+        Ok(Self {
+            transport,
+            rx,
+            tx,
+            event,
+            guest_cid,
+            queue_buf_dma,
+            queue_buf_rx,
+        })
+    }
+
+    /// Connect to the destination.
+    pub fn connect(&mut self, dst_cid: u64, src_port: u32, dst_port: u32) -> Result {
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: dst_cid.into(),
+            src_port: src_port.into(),
+            dst_port: dst_port.into(),
+            op: Op::VIRTIO_VSOCK_OP_REQUEST.into(),
+            ..Default::default()
+        };
+        self.tx
+            .add_notify_wait_pop(&[header.as_bytes(), &[]], &mut [], &mut self.transport)?;
+        let token = if let Some(token) = self.rx.peek_used() {
+            token // TODO: Use let else after updating Rust
+        } else {
+            return Err(SocketError::NoResponseReceived.into());
+        };
+        let _len = unsafe {
+            self.rx
+                .pop_used(token, &[], &mut [&mut self.queue_buf_rx])?
+        };
+        let packet_rx = VirtioVsockPacket::read_from(&self.queue_buf_rx)?;
+        let result = match packet_rx.op()? {
+            Op::VIRTIO_VSOCK_OP_RESPONSE => Ok(()),
+            Op::VIRTIO_VSOCK_OP_RST => Err(SocketError::ConnectionFailed.into()),
+            Op::VIRTIO_VSOCK_OP_INVALID => Err(SocketError::InvalidOperation.into()),
+            _ => todo!(),
+        };
+        if result.is_err() {
```

Comment:
This seems like it should be the caller's responsibility to log if it wants to. If you want to log extra details here for debugging, how about just adding a `trace!("Received packet {:?}", packet_rx);` before the match? Then you don't need `result` either, the match can be the return value of the method.